src/transformers/models/qwen3/modular_qwen3.py

```diff
@@ -0,0 +1,148 @@
+from typing import Callable, Optional, Tuple
+
+import torch
+import torch.utils.checkpoint
+from torch import nn
+
+from ...cache_utils import Cache
+from ...modeling_flash_attention_utils import FlashAttentionKwargs
+from ...modeling_utils import ALL_ATTENTION_FUNCTIONS
+from ...processing_utils import Unpack
+from ...utils import logging
+from ..llama.modeling_llama import (
+    LlamaAttention,
+    LlamaDecoderLayer,
+    LlamaForCausalLM,
+    LlamaForQuestionAnswering,
+    LlamaForSequenceClassification,
+    LlamaForTokenClassification,
+    LlamaMLP,
+    LlamaRMSNorm,
+    apply_rotary_pos_emb,
+    eager_attention_forward,
+)
+from ..mistral.modeling_mistral import MistralModel
+from .configuration_qwen3 import Qwen3Config
+
+
+logger = logging.get_logger(__name__)
+
+
+class Qwen3RMSNorm(LlamaRMSNorm):
+    pass
+
+
+class Qwen3MLP(LlamaMLP):
+    def __init__(self, config):
+        super().__init__(config)
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+
+
+class Qwen3Attention(LlamaAttention):
```

Comment:
```suggestion
class Qwen3Attention(Olmo2Attention):
```
Olmo2 already incorporates the q / k norm!

Reply:
Qwen3's q/k norm is shared across attention heads, which is slightly different from Olmo2.

Reply:
Ok!

Reply:
BTW the fact that it is shared is not shown / implemented here. I suppose the weights are replicated? if so We can still use Olmo2 😉 unless we rename them as `shared_q_k_norm`